storage: c++ warnings when compiling db.cc

using `cc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609`

```
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In function ‘bool SplitKey(rocksdb::Slice, rocksdb::Slice*, rocksdb::Slice*)’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:248:15: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if (ts_size >= buf.size()) {
               ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In member function ‘virtual bool {anonymous}::DBMergeOperator::FullMerge(const rocksdb::Slice&, co
nst rocksdb::Slice*, const std::deque<std::__cxx11::basic_string<char> >&, std::__cxx11::string*, rocksdb::Logger*) const’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:782:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int i = 0; i < operand_list.size(); i++) {
                       ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In member function ‘virtual bool {anonymous}::DBMergeOperator::PartialMergeMulti(const rocksdb::Sl
ice&, const std::deque<rocksdb::Slice>&, std::__cxx11::string*, rocksdb::Logger*) const’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:802:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int i = 0; i < operand_list.size(); i++) {
                       ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In member function ‘virtual void {anonymous}::DBLogger::Logv(const char*, __va_list_tag*)’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:852:34: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if ((result >= 0) && (result < sizeof(space))) {
                                  ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In member function ‘DBSSTable* DBEngine::GetSSTables(int*)’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:1344:21: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for (int i = 0; i < metadata.size(); i++) {
                     ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:1354:65: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-alia
sing]
       tables[i].start_key.key = *reinterpret_cast<DBSlice*>(&str);
                                                                 ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:1354:31: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-alia
sing]
       tables[i].start_key.key = *reinterpret_cast<DBSlice*>(&str);
                               ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:1359:63: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-alia
sing]
       tables[i].end_key.key = *reinterpret_cast<DBSlice*>(&str);
                                                               ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:1359:29: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-alia
sing]
       tables[i].end_key.key = *reinterpret_cast<DBSlice*>(&str);
                             ^
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc: In function ‘MVCCStatsResult MVCCComputeStatsInternal(rocksdb::Iterator*, DBKey, DBKey, int64_t)’:
/mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/db.cc:2230:30: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         if (meta.val_bytes() != value.size()) {
```

--- a/pkg/ccl/storageccl/engineccl/rocksdb.go
+++ b/pkg/ccl/storageccl/engineccl/rocksdb.go
@@ -18,7 +18,7 @@ import (
 
 // #cgo CPPFLAGS: -I../../../../vendor/github.com/cockroachdb/c-protobuf/internal/src
 // #cgo CPPFLAGS: -I../../../../vendor/github.com/cockroachdb/c-rocksdb/internal/include
-// #cgo CXXFLAGS: -std=c++11 -Wall
+// #cgo CXXFLAGS: -std=c++11 -Werror -Wall -Wno-sign-compare
 // #cgo !strictld,darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
 // #cgo !strictld,!darwin LDFLAGS: -Wl,-unresolved-symbols=ignore-all
 // #cgo linux LDFLAGS: -lrt

--- a/pkg/storage/engine/db.cc
+++ b/pkg/storage/engine/db.cc
@@ -1351,12 +1351,12 @@ DBSSTable* DBEngine::GetSSTables(int* n) {
       // This is a bit ugly because we want DBKey.key to be copied and
       // not refer to the memory in metadata[i].smallestkey.
       DBString str = ToDBString(tmp);
-      tables[i].start_key.key = *reinterpret_cast<DBSlice*>(&str);
+      tables[i].start_key.key = DBSlice{str.data, str.len};
     }
     if (DecodeKey(metadata[i].largestkey, &tmp,
                   &tables[i].end_key.wall_time, &tables[i].end_key.logical)) {
       DBString str = ToDBString(tmp);
-      tables[i].end_key.key = *reinterpret_cast<DBSlice*>(&str);
+      tables[i].end_key.key = DBSlice{str.data, str.len};
     }
   }
   return tables;

--- a/pkg/storage/engine/rocksdb.go
+++ b/pkg/storage/engine/rocksdb.go
@@ -51,7 +51,7 @@ import (
 
 // #cgo CPPFLAGS: -I../../../vendor/github.com/cockroachdb/c-protobuf/internal/src
 // #cgo CPPFLAGS: -I../../../vendor/github.com/cockroachdb/c-rocksdb/internal/include
-// #cgo CXXFLAGS: -std=c++11 -Wall
+// #cgo CXXFLAGS: -std=c++11 -Werror -Wall -Wno-sign-compare
 // #cgo linux LDFLAGS: -lrt
 //
 // #include <stdlib.h>

